KEP-2401: Kubeflow LLM Trainer V2 (#2410)
kubeflow/training-operator: docs/proposals/2401-llm-trainer-v2/README.md

# KEP-2401: Kubeflow LLM Trainer V2

## Authors

- Shao Wang - [@Electronic-Waste](https://github.com/Electronic-Waste)

Creation date: 2025-01-31

Google doc: http://bit.ly/4gp8JGd

## Overview

> This document discusses the design of LLM Trainer for [Kubeflow Training v2](../2170-kubeflow-training-v2/README.md), tracked by [this issue](https://github.com/kubeflow/training-operator/issues/2401).

Author: SGTM.

(The following picture comes from the slides of [Andrey and Yuki’s talk in KubeCon NA 2024](https://kccncna2024.sched.com/event/1i7nV?iframe=no), which explains the workflow of LLM fine-tuning very well)

![](./llm-fine-tuning-lifecycle.png)

## Motivation

Fine-tuning LLMs on Kubernetes is challenging for Data Scientists due to the complex Kubernetes configurations, diverse fine-tuning techniques, and different distributed strategies like data and model-parallelism. It’s crucial to hide the complex infrastructure configurations from users, and allow them to gracefully shift among diverse fine-tuning techniques and distributed strategies.

By now, Kubeflow Training V1 has implemented a [Trainer for LLM](../2003-train-api/README.md) based on the HuggingFace library. However, it is **integrated with data and model initialization**, which are separated in Kubeflow Training V2, and has **limited fine-tuning mechanism support**. We need to migrate it to V2 and support more fine-tuning configurations and distributed strategies in accordance with the growing needs in custom LLM fine-tuning.

### Goals

> - Build LLM Trainer V2 based on `torchtune`.

Reviewer: @Electronic-Waste What do you think about naming this as:
Build LLM Fine-Tuning recipes using the `torchtune` and Kubeflow Trainer ClusterTrainingRuntime?
Since we don't actually build custom LLM Trainer, maybe it is clearer ?

> - Update Kubeflow Training SDK to fine-tune LLMs flexibly.

Author: Better than my expression! 👍

INFO:torchtune.utils.logging:Optimizer and loss are initialized.
INFO:torchtune.utils.logging:Loss is initialized.
INFO:torchtune.utils.logging:Dataset and Sampler are initialized.
INFO:torchtune.utils.logging:Learning rate scheduler is initialized.
1|52|Loss: 2.3697006702423096:   0%|▏                     | 52/25880 [00:24<3:55:01,  1.83it/s]
```

By adopting `torchtune` as the low-level runtime for LLM fine-tuning, we can easily obtain the flexibility, efficiency and scalability brought by its unique "recipe-config" design, which will surely streamline and scale LLM fine-tuning on Kubernetes.

To hide users from complex Kubernetes configuations, we will provide a simple yet flexible Python SDK wrapping all specifications of models, datasets, training runtime and fine-tuning configs. Like this:

```python
job_id = TrainingClient().train(
    dataset_config=HuggingFaceDatasetConfig(
        storage_uri="tatsu-lab/alpaca",
    ),
    trainer=Trainer(
        fine_tuning_config=FineTuningConfig(
            peft_config=LoraConfig(r=4),
            dtype="bf16",
        ),
        num_nodes=5,
    ),
    runtime_ref=llm_runtime,
)
```

> ## Design Details

Reviewer: I think, we should have a section about Datasets and their configurations.
Do we have any requirements for DS format with torchtune?
E.g. Instruct, Chat, Multimodal datasets.

Author: `torchtune` has already supported some Dataset classes: https://pytorch.org/torchtune/main/basics/datasets_overview.html

I think, we may reuse `torchtune`'s preprocessing logic after we download the dataset with dataset initializer.

Reviewer: Does it perform tokenization automatically before FT ?

### `torchtune` Plugin

> As is shown in the [torchtune official document](https://pytorch.org/torchtune/main/tune_cli.html#run-a-recipe) and [source code](https://github.com/pytorch/torchtune/blob/75965d4281b9b76c454630d015221b9933c77bf3/torchtune/_cli/run.py#L113-L118), the distributed training arguments like `--nnodes` and `--nproc_per_node` should be passed ahead of the recipe argument in the command line, and **cannot be passed by the environment variables** in the `PET_XXX` convention. And also, `torchtune` is extremely different from the fine-tuning paradigm of `torchrun` because it is **recipe and config-based**, which may need more mutation operations in the config file. Here is an [example](https://github.com/Electronic-Waste/kubeflow-llm-trainer/blob/main/torchtune-llm-finetuning.yaml).

Reviewer: Did you actually try to use `PET_xxx` envs with `torchtune` ?
@joecummings Does the `torchtune` CLI explicitly wait for users to write `--nnodes` and `--nproc_per_node` args ?

Author: I've searched the source code and docs, and found that we can only pass these parameters with CLI args.

Reviewer: @ebsmothers Is there any way for us to allow users leverage `PET_` envs for `torch tune` CLI also ?
As of now, it reads [them only from the CLI](https://github.com/pytorch/torchtune/blob/e6cba2532d51a53936c7646bd4cdaa6b2b57ed66/torchtune/_cli/run.py#L113-L117).

However, [the `torchrun` CLI](https://github.com/pytorch/pytorch/blob/main/torch/distributed/run.py#L398) can read it from [the `PET_` env.](https://github.com/pytorch/pytorch/blob/main/torch/distributed/argparse_util.py#L44-L56)

Reviewer 2: I think this might require modifying our args parser a bit. I notice that we use `_add_action` on our `tune run` subparser [here](https://github.com/pytorch/torchtune/blob/e6cba2532d51a53936c7646bd4cdaa6b2b57ed66/torchtune/_cli/run.py#L80) to pass the torchrun args through, I wonder whether we can instead use `add_argument` with torchrun's `env` Action? (Admittedly my argparse knowledge is a bit rough, @joecummings will definitely have a better idea here.)

Author: @andreyvelich I think it's okay to pass arguments in the CLI for `torchtune`. We just need to add another runtime plugin for this and handle the mutating logic separately, which will also help us decompose the handling logic of LLM fine-tuning (launched by `torchtune`) from PyTorch DDP (launched by `torchrun`). It's more flexible and will surely make our implementation neat and modularized.

We can embed `TorchMLPolicySource` into `TorchtuneMLPolicySource`:

```go
// MLPolicySource represents the runtime-specific configuration for various technologies.
// One of the following specs can be set.
type MLPolicySource struct {
    // Configuration for the PyTorch runtime.
    Torch *TorchMLPolicySource `json:"torch,omitempty"`

    // Configuration for the Torchtune runtime.
    Torchtune *TorchtuneMLPolicySource `json:"torchtune,omitempty"`

    // Configuration for the MPI Runtime.
    MPI *MPIMLPolicySource `json:"mpi,omitempty"`
}

// TorchtuneMLPolicySource represents a Torchtune runtime configuration.
// Indicate to use Torchtune runtime plugin instead of PyTorch Plugin.
type TorchtuneMLPolicySource struct {
    // Configuration for the PyTorch runtime
    *TorchMLPolicySource `json:",inline"`
}
``` 

And in the future, we may add more parameters to `TorchtuneMLPolicy` according to our need. Isn't it great?

Author: SGTM. I'll update the doc.

Thus, we need to implement a new plugin for `torchtune` if we decide to adopt `torchtune` as a launcher for LLM fine-tuning on Kubernetes. And the new plugin should have these abilities:

1. Parse distributed training arguments in TrainJob and TrainingRuntime API, and integrate them with the `tune run` command.

> 2. Handle overrides in the `torchtune` fine-tuning configuration file.

Reviewer: I think, we should have a dedicated section on how we will perform overrides.
So we can explain how we are going to propagate values from Kubeflow SDK into `torchtune` recipes.

Author: Yeah, I think so. I'm still updating the doc. Hopefully, it will come out before the community meeting:)

3. Validate some requirements.

> ### Fine-Tuning Config

Reviewer: What about other FT args, like [num epochs, lr, loss, etc.](https://github.com/pytorch/torchtune/blob/75965d4281b9b76c454630d015221b9933c77bf3/recipes/configs/qwen2/7B_full.yaml#L54-L67)?
Are we going to gracefully add them in the future, once users request it ?

Author: I'm still updating the doc. It will come out soon:)

We will add the fine-tuning configurations in the `fine_tuning_config` field in `Trainer` dataclass.

| Parameters | What is it? |
| - | - |

> | dtype | The underlying data type used to represent the model and optimizer parameters. Currently, we support `bf16` and `fp32`. |

Reviewer: What about fp16 ?
Do we know if all `torchtune` recipes support these dtypes ?

Author: `touchtune` only supports `bf16` and `fp32`: https://pytorch.org/torchtune/main/tutorials/memory_optimizations.html#model-precision

| peft_config | Configuration for the PEFT(Parameter-Efficient Fine-Tuning), including Lora, AdapterPrompt, PrefixTuning, etc. |

```python
# FineTuningConfig DataClass
@dataclass

> class FineTuningConfig:

Reviewer: Do you prefer to keep `torchtune` parameters under FineTuningConfig or create a new dataclass: TorchTuneConfig which will be specific to `torchtune` ?
I am asking because in the future if we support other LLM Trainers (e.g. Nvidia NeMO) it might not support all parameters. cc @deepanker13

Author: Yeah! It's a good idea. I prefer creating a new dataclass and has two options:

Option1: Nest `TorchtuneConfig` into `FineTuningConfig`

```python
@dataclass
class FineTuningConfig:
    torchtune_config: Optional[TorchtuneConfig] = None,
    nemo_config: Optional[NemoConfig] = None,

```

Option2: Discard `FineTuningConfig`

```python
@dataclass
class Trainer:
    """Trainer configuration.
    TODO: Add the description
    """

    func: Optional[Callable] = None
    func_args: Optional[Dict] = None
    packages_to_install: Optional[List[str]] = None
    pip_index_url: str = constants.DEFAULT_PIP_INDEX_URL
    fine_tuning_config: Optional[Union[TorchtuneConfig,NemoConfig]] = None
    num_nodes: Optional[int] = None
    resources_per_node: Optional[dict] = None
```

Which one do you prefer?
/cc @kubeflow/wg-training-leads @franciscojavierarceo @astefanutti @deepanker13

Reviewer: Another option that came to my mind is to separate Trainer and FT Config since they are mutually exclusive :
```py
def train(
        self,
        runtime_ref: str,
        trainer: Optional[types.Trainer] = None,
        fine_tune_config: Optional[types.FineTuningConfig] = None,
        dataset_config: Optional[types.HuggingFaceDatasetConfig] = None,
        model_config: Optional[types.HuggingFaceModelInputConfig] = None,
    ) -> str:
     pass

@dataclass
class TorchTuneConfig:
    dtype = "bf32"
    peft_config: Optional[LoraConfig] = None

@dataclass
class NeMoConfig:
    lr = 0.01

@dataclass
class FineTuningConfig:
    torchtune_config: Optional[TorchTuneConfig] = None
    nemo_config: Optional[NeMoConfig] = None
    num_nodes: Optional[int] = None
    resources_per_node: Optional[dict] = None

    def __post_init__(self):
        if self.torchtune_config and self.nemo_config:
            raise ValueError("Only one of `torchtune_config` or `nemo_config` can be set.")
 
 ```
 
 Example of using:
 
 
 ```python
 TrainerClient().train(
    fine_tune_config=FineTuningConfig(
        torch_tune_config=TorchTuneConfig(
            dtype="fp32",
        ),
        num_nodes=5,
    ),
)

 ```
 
 But I am not a perfect Python expert 😄 
 
 @droctothorpe Any thoughts ?

Author: I prefer placing `fine_tuning_config` in `trainer` field, since `trainer` serves as the wrapper for training parameters (from my perspective). And personally, I prefer option2.

Also, how will you handle the `num_nodes` and `resources_per_node` field in `trainer`? Won't distributed training such as PyTorch DDP need it?

Author: Inspired by https://github.com/kubeflow/trainer/issues/2430, another idea based on the lastest comment is to call:
- `custom_training_config` -> `pre_training_config`
- `fine_tuning_config` -> `post_training_config`

Reviewer: One of the idea is to put: `num_nodes` and `resources_per_node` under `scalingConfig` Data class.

>It's better to put fine_tuning_config into the trainer field to be more consistent with the fields we defined in TrainJob

To be honest, we don't really need to be 100% consistent with TrainJob since Python APIs should be ML User focused.

>Inspired by https://github.com/kubeflow/trainer/issues/2430, another idea based on the lastest comment is to call:

I don't think we should do it, since user can specify fine-tuning script using the Training Function, as I described in #2430.

---

In general, let's think about naming when users:

- **Manually define a training function** – A self-contained function that encapsulates the entire model training process.
- **Configure parameters for a pre-existing LLM Trainer** – A trainer that already includes fine-tuning logic, requiring only parameter adjustments.

Reviewer: cc @saileshd1402 @johnugeorge for more ideas

Author: > To be honest, we don't really need to be 100% consistent with TrainJob since Python APIs should be ML User focused.

SGTM. Then how about defining:

- **Manually define a training function**: Name it as `custom_training_config` in the `train` API.
- **Configure parameters for a pre-existing LLM Trainer**: Name it as `trainer` in the `train` API. `trainer` means that we've already defined the training/fine-tuning logic, just like HF's Trainer and `torchtune`. It's not limited to pre-training or post-training. We just define it according to the training pattern.

```python
def train(
    self,
    runtime_ref: str,
    custom_training_config: Optional[CustomTrainingConifg] = None,
    trainer: Optional[Union[TorchtuneConfig]] = None,
    scaling_config: Optional[ScalingConfig] = None,
    dataset_config: Optional[types.HuggingFaceDatasetConfig] = None,
    model_config: Optional[types.HuggingFaceModelInputConfig] = None,
) -> str:
    pass

@dataclass
class ScalingConfig
    num_nodes: Optional[int] = None
    proc_per_node: Optional[int] = None
    resources_per_node: Optional[int] = None

@dataclass
class CustomTrainingConfig
    func: Optional[Callable] = None
    func_args: Optional[Dict] = None
    packages_to_install: Optional[List[str]] = None
    pip_index_url: str = constants.DEFAULT_PIP_INDEX_URL

@dataclass
class TorchtuneConfig:
    recipe: str
    config: str
    device: Optional[str] = None
    dtype: Optional[str] = None
    batch_size: Optional[int] = None
    epochs: Optional[int] = None
    loss: Optional[str] = None
    optimizer_config: Optional[OptimizerConfig] = None
    scheduler_config: Optional[SchedulerConfig] = None
    gradient_accumulation_steps: Optional[int] = None
    enable_activation_checkpointing: Optional[bool] = False
    enable_activation_offloading: Optional[bool] = False
    peft_config: Optional[Union[LoraConfig]] = None
    dataset_preprocess_config: Optional[
        Union[TorchtuneInstructDataset, TorchtuneChatDataset, TorchtuneMultimodalDataset],
    ] = None

```

/cc @kubeflow/wg-training-leads @franciscojavierarceo @astefanutti @deepanker13

    dtype: string = "bf16"
    peft_config: Optional[Union[LoraConfig]] = None

```

The Python SDK will look like:

**LoRA Config**

The *LoraConfig* represents the config of LoRA we use to fine-tune the model.

> | Parameters | What is it? |

Reviewer: Let's also add the type of the parameter (e.g. list, str)

Author: I'll add them in the following dataclass definition. I'll upload the updated docs later:)

| - | - |
| apply_lora_to_mlp | Whether to apply LoRA to the MLP in each transformer layer |
| apply_lora_to_output | Whether to apply LoRA to the model’s final output projection |
| lora_attn_modules | A list of strings specifying which layers of the model to apply LoRA: 1. `q_proj` applies LoRA to the query projection layer. 2. `k_proj` applies LoRA to the key projection layer. 3. `v_proj` applies LoRA to the value projection layer. 4. `output_proj` applies LoRA to the attention output projection layer. |
| lora_rank | The rank of the low rank decomposition. |
| lora_alpha | The scaling factor that adjusts the magnitude of the low-rank matrices’ output |
| lora_dropout | The probability of applying Dropout to the low rank updates |
| quantize_base | Whether to enable model quantization |

> | use_dora | Whether to enable DoRA |

Reviewer: Is quantization supported for all LLMs ?

Author: For every LLM that support LoRA, it also supports QLoRA:

```
lora_finetune_single_device              llama2/7B_lora_single_device            
                                         llama2/7B_qlora_single_device           
                                         code_llama2/7B_lora_single_device       
                                         code_llama2/7B_qlora_single_device      
                                         llama3/8B_lora_single_device            
                                         llama3_1/8B_lora_single_device          
                                         llama3/8B_qlora_single_device           
                                         llama3_2/1B_lora_single_device          
                                         llama3_2/3B_lora_single_device          
                                         llama3/8B_dora_single_device            
                                         llama3/8B_qdora_single_device           
                                         llama3_1/8B_qlora_single_device         
                                         llama3_2/1B_qlora_single_device         
                                         llama3_2/3B_qlora_single_device         
                                         llama2/13B_qlora_single_device          
                                         mistral/7B_lora_single_device           
                                         mistral/7B_qlora_single_device          
                                         gemma/2B_lora_single_device             
                                         gemma/2B_qlora_single_device            
                                         gemma/7B_lora_single_device             
                                         gemma/7B_qlora_single_device            
                                         gemma2/2B_lora_single_device            
                                         gemma2/2B_qlora_single_device           
                                         gemma2/9B_lora_single_device            
                                         gemma2/9B_qlora_single_device           
                                         gemma2/27B_lora_single_device           
                                         gemma2/27B_qlora_single_device          
                                         phi3/mini_lora_single_device            
                                         phi3/mini_qlora_single_device           
                                         qwen2/7B_lora_single_device             
                                         qwen2/0.5B_lora_single_device           
                                         qwen2/1.5B_lora_single_device           
                                         qwen2_5/0.5B_lora_single_device         
                                         qwen2_5/1.5B_lora_single_device         
                                         qwen2_5/3B_lora_single_device           
                                         qwen2_5/7B_lora_single_device           
                                         qwen2_5/14B_lora_single_device          
                                         llama3_2_vision/11B_lora_single_device  
                                         llama3_2_vision/11B_qlora_single_device
lora_finetune_distributed                llama2/7B_lora                          
                                         llama2/13B_lora                         
                                         llama2/70B_lora                         
                                         llama2/7B_qlora                         
                                         llama2/70B_qlora                        
                                         llama3/8B_dora                          
                                         llama3/70B_lora                         
                                         llama3_1/70B_lora                       
                                         llama3_3/70B_lora                       
                                         llama3_3/70B_qlora                      
                                         llama3/8B_lora                          
                                         llama3_1/8B_lora                        
                                         llama3_2/1B_lora                        
                                         llama3_2/3B_lora                        
                                         llama3_1/405B_qlora                     
                                         mistral/7B_lora                         
                                         gemma/2B_lora                           
                                         gemma/7B_lora                           
                                         gemma2/2B_lora                          
                                         gemma2/9B_lora                          
                                         gemma2/27B_lora                         
                                         phi3/mini_lora                          
                                         qwen2/7B_lora                           
                                         qwen2/0.5B_lora                         
                                         qwen2/1.5B_lora                         
                                         qwen2_5/0.5B_lora                       
                                         qwen2_5/1.5B_lora                       
                                         qwen2_5/3B_lora                         
                                         qwen2_5/7B_lora                         
                                         qwen2_5/32B_lora                        
                                         qwen2_5/72B_lora                        
                                         llama3_2_vision/11B_lora                
                                         llama3_2_vision/11B_qlora               
                                         llama3_2_vision/90B_lora                
                                         llama3_2_vision/90B_qlora
```

And currently, we do not support DPO, which does not support QLoRA:

```
lora_dpo_single_device                   llama2/7B_lora_dpo_single_device        
                                         llama3_1/8B_lora_dpo_single_device      
lora_dpo_distributed                     llama2/7B_lora_dpo                      
                                         llama3_1/8B_lora_dpo
```

> 

Reviewer: It would be nice if you could also add the section of how we are going to structure these recipes.
And how we are going to select the default resources, number of PyTorch nodes.

## Implementation History

> 

Reviewer: Let's also have a section about model and dataset initialization.
Can we leverage the existing Kubeflow Trainer Initializers: https://github.com/kubeflow/trainer/blob/master/pkg/initializer/dataset/huggingface.py ?
cc @akshaychitneni @shravan-achar

Author: Will add them soon. Still updating the doc:)

1. PEFT(Parameter Efficient Fine Tuning): Including LoRA, QLoRA, Adapter Prompt and Prefix Tuning.

2. Sharding Policies: Including FSDP and DeepSeed ZeRO.

The workflow of LLM Trainer V2 is shown in the graph below:

![](./llm-trainer-v2-workflow.png)

To hide users from complex Kubernetes configuations, we will provide a simple yet flexible Python SDK wrapping all specifications of models, datasets, training runtime and fine-tuning configs. Like this:

```python
TrainingClient().train(
    dataset_config=HuggingFaceDatasetConfig(
        storage_uri="tatsu-lab/alpaca",
    ),
    trainer=Trainer(
        fine_tuning_config=FineTuningConfig(
            framework="huggingface",
            dataset_class="Instruction",
            peft_config=LoraConfig(r=4),
            sharding_config=FsdpConfig(...),
            kwargs={},
        ),
        num_nodes=5,
    ),
    runtime_ref=llm_runtime,
)
```

And it's worthwhile to notice that we'll preprocess dataset for users with builtin dataset classes or a customized one. If users want to preprocess datasets by themselves, they need to implement a customized data class with specified methods implemented and pass it to the Python SDK.

> In the future, we'll provide users with more options on launchers (`torchtune`, `accelerate`), frameworks (TensorFlow, Jax, etc.) and fine-tuning techniques (RLHF, Distilation, etc.).

Reviewer 3: torchtune would be convenient for Llama-stack https://github.com/meta-llama/llama-stack/blob/529708215c5ad54e1ef41ba3e68d3a2af8d563b0/llama_stack/providers/inline/post_training/torchtune/post_training.py#L32

Reviewer 3: Oh, that's unfortunate.

Reviewer 4: Hey @Electronic-Waste and @franciscojavierarceo, we just merged official multi-node support in https://github.com/pytorch/torchtune/pull/2301. It's available in nightly builds and builds from source right now, but will be released in the next week or so in our v0.6 version.

Author: @joecummings Thanks for updating this. It's a really big news for us!

I noticed that in https://github.com/pytorch/torchtune/pull/2348 the README was updated to add multinode support, but only full finetuning supports multinode mode: https://github.com/pytorch/torchtune#finetuning-recipes. May I ask whether torchtune has plans to support multinode mode in other finetuning methods in the future?

> I think, if we initially add Kubernetes blueprints for LLMs that torchtune supports, we should be good: https://github.com/pytorch/torchtune/tree/main?tab=readme-ov-file#models.

@andreyvelich SGTM.

Reviewer 4: Yep, we'll have multi node support for all our distributed recipes soon!

Technically, multi node is already supported in all our distributed recipes. The reason we don't claim support is b/c we haven't officially benchmarked it + we want to land other optimizations like initialization of the process group on GPU if available and 2D parallelism.

| Parameters | What is it? |
| - | - |
| num_virtual_tokens | The number of virtual tokens |

```python
# PrefixConfig DataClass
@dataclass
class PrefixConfig:
    num_virtual_tokens: int = 30

```

#### FSDP Config

The *FsdpConfig* represents the config of FSDP we use to fine-tune the model.

| Parameters | What is it? |
| - | - |
| mixed_precision | Whether to enable mixed precision training |
| use_fp16 | Whether to use FP16 during the mixed precision training |
| fsdp_cpu_offload | Whether to offload some weights and optimizer states to cpu |
| sharding_strategy | The sharding strategy for FSDP, e.g. FULL_SHARD (default), HYBRID_SHARD, SHARD_GRAD_OP, NO_SHARD. |
| hsdp | Whether to enable Hybrid Shard Data Parallel (HSDP) |
| sharding_group_size | Specify the GPU num in the sharding group when hsdp set to true |
| replica_group_size | The number of sharding groups |
| checkpoint_type | Specify the type of model checkpoints |
| fsdp_activation_checkpointing | Whether to enable Activation Checkpointing |

```python
# FsdpConfig DataClass
@dataclass
class FsdpConfig:
    mixed_precision: bool = True
    use_fp16: bool = False
    fsdp_cpu_offload: bool=False

>     sharding_strategy: ShardingStrategy = ShardingStrategy.FULL_SHARD

Reviewer 3: Amazing, thanks @joecummings!

>     sharding_strategy: ShardingStrategy = ShardingStrategy.FULL_SHARD

Author: Thanks for the info @joecummings!

    hsdp: bool = False
    sharding_group_size: int = 0 # requires hsdp to be set.
    replica_group_size: int = 0 #requires hsdp to be set.
    checkpoint_type: StateDictType = StateDictType.SHARDED_STATE_DICT

>     fsdp_activation_checkpointing: bool = True

Reviewer 4: Cool opportunity here to support activation offloading, too, which @janeyx99 was nice enough to develop and implement in torchtune! https://github.com/pytorch/torchtune/pull/1443

Author: @joecummings That's amazing!

                                         mistral/7B_full_low_memory
full_finetune_distributed                llama2/7B_full
                                         llama2/13B_full
                                         mistral/7B_full
lora_finetune_single_device              llama2/7B_lora_single_device
                                         llama2/7B_qlora_single_device
                                         mistral/7B_lora_single_device

$ tune run lora_finetune_single_device --config llama2/7B_lora_single_device epochs=1
INFO:torchtune.utils.logging:Running LoRAFinetuneRecipeSingleDevice with resolved config:
Writing logs to /tmp/lora_finetune_output/log_1713194212.txt
INFO:torchtune.utils.logging:Model is initialized with precision torch.bfloat16.
INFO:torchtune.utils.logging:Tokenizer is initialized from file.
INFO:torchtune.utils.logging:Optimizer and loss are initialized.
INFO:torchtune.utils.logging:Loss is initialized.
INFO:torchtune.utils.logging:Dataset and Sampler are initialized.
INFO:torchtune.utils.logging:Learning rate scheduler is initialized.
1|52|Loss: 2.3697006702423096:   0%|▏                     | 52/25880 [00:24<3:55:01,  1.83it/s]
```

> (**Note**: We need to create a new plugin for `torchtune`, so that it can fit in the yaml-based fine-tuning configurations. And also we may need to explore how to integrate the recipes provided by `torchtune`.)

Reviewer: Absolutely, @joecummings!
Do you have any community calls where we can discuss it further ?
If you are available, we can talk about it during the [next Kubeflow Training Working Group](https://docs.google.com/document/d/1MChKfzrKAeFRtYqypFbMXL6ZIc_OgijjkvbqmwRV-64/edit?tab=t.0#heading=h.9ki8vcje7ewu) call on **Feb 19th at 5pm UTC.**

> (**Note**: We need to create a new plugin for `torchtune`, so that it can fit in the yaml-based fine-tuning configurations. And also we may need to explore how to integrate the recipes provided by `torchtune`.)

Reviewer 4: Ahh, unfortunate timing - we just had a call last Thursday. More than willing to find some additional time for us to meet, but I'm also free to join your meeting on February 19th so we can talk more then :)

INFO:torchtune.utils.logging:Running LoRAFinetuneRecipeSingleDevice with resolved config:
Writing logs to /tmp/lora_finetune_output/log_1713194212.txt
INFO:torchtune.utils.logging:Model is initialized with precision torch.bfloat16.
INFO:torchtune.utils.logging:Tokenizer is initialized from file.
INFO:torchtune.utils.logging:Optimizer and loss are initialized.
INFO:torchtune.utils.logging:Loss is initialized.
INFO:torchtune.utils.logging:Dataset and Sampler are initialized.
INFO:torchtune.utils.logging:Learning rate scheduler is initialized.
1|52|Loss: 2.3697006702423096:   0%|▏                     | 52/25880 [00:24<3:55:01,  1.83it/s]
```

By adopting `torchtune` as the low-level runtime for LLM fine-tuning, we can easily obtain the flexibility, efficiency and scalability brought by its unique "recipe-config" design, which will surely streamline and scale LLM fine-tuning on Kubernetes.

To hide users from complex Kubernetes configuations, we will provide a simple yet flexible Python SDK wrapping all specifications of models, datasets, training runtime and fine-tuning configs. Like this:

```python

> job_id = TrainingClient().train(

Author: Thanks for pointing this out!

tune run --nnodes=1 --nproc-per-node=4 lora_finetune_distributed \
    --config llama3/8B_lora \
    model.lora_attn_modules=[q_proj,k_proj,v_proj] \
    model.apply_lora_to_mlp=True \
    model.lora_rank=64 \
    model.lora_alpha=128

```

The CLI-based way allows us to mutate these parameters by overriding the `command` and `args` fields in the Trainer Node. By specifying override parameters in the TrainJob `.spec.trainer.command` and `.spec.trainer.args` fields, we can easily mutate these parameters, just like:

```YAML
apiVersion: kubeflow.org/v2alpha1
kind: ClusterTrainingRuntime
metadata:
  name: torchtune-llm-finetuning
  labels:
    training.kubeflow.org/phase: post-training
spec:
  mlPolicy:
    numNodes: 1
    torch:
      numProcPerNode: auto
  template:
    spec:
      replicatedJobs:
        - name: Node
          template:
            spec:
              template:
                spec:
                  containers:
                    - name: trainer
                      image: <pytorch+cuda+torchtune image>
                      command:

>                         - tune ls

Reviewer: I think, by default our runtime should be able to fine-tune model without any additional configurations from users.
So user can do something like this:
```py
TrainerClient().train(
  runtime_ref="torchtune-llama-3.3-70b"
)

In that case, we will just use the default settings that being configured under runtime.

Reviewer 3: +1

Author: > I think, by default our runtime should be able to fine-tune model without any additional configurations from users.

That's ideal. But we could hardly achieve this because:

1. **Config files are different across different fine-tuning techniques**: As I mentioned here, they cannot share config files. So we cannot only use one ClusterTrainingRuntime for a model if we do not use `recipe` and `config` parameters in . https://github.com/kubeflow/trainer/pull/2410#discussion_r1970153947 https://github.com/kubeflow/trainer/pull/2410#discussion_r1971693997
2. **Explosion in the number of ClusterTrainingRuntimes**: Since we can't share config files, it's unimaginable for us to maintain ~100 ClusterTrainingRuntimes https://github.com/kubeflow/trainer/pull/2410#discussion_r1970153947

>                         - tune ls

Reviewer: Can we wrap the default torchtune config under TrainingRuntime args, and override the user values using `tune run` arguments ?
So the experience will be:
1. ClusterTrainingRuntime contains the default torchtune config to fine-tune model.
2. User can override values using `TorchTuneConfig()` class.

        // which will be inserted in front of Args of Trainer Node.
        // TBA
    }
    // ...
}
```

Related changes:

1. Add `Args` fields to Info (`pkg/runtime/runtime.go`).

```go
type Trainer struct {
    NumNodes       *int32
    NumProcPerNode string
    Env           []corev1.EnvVar
    Args          []string
    ContainerPort *corev1.ContainerPort
    Volumes       []corev1.Volume
    VolumeMounts  []corev1.VolumeMount
}
```

2. Insert `Info.Args` in front of Args of Trainer Node (`pkg/runtime/framework/plugins/jobest/builder.go`)

```go
// ...

// Update the Trainer args.
if info.Trainer.Args != nil {
    var args []string
    copy(args, info.Trainer.Args)
    b.Spec.ReplicatedJobs[i].Template.Spec.Template.Spec.Containers[j].Args = append(
        args, b.Spec.ReplicatedJobs[i].Template.Spec.Template.Spec.Containers[j].Ports...)
}

// Update the Trainer container port.
if info.Trainer.ContainerPort != nil {
    b.Spec.ReplicatedJobs[i].Template.Spec.Template.Spec.Containers[j].Ports = append(
        b.Spec.ReplicatedJobs[i].Template.Spec.Template.Spec.Containers[j].Ports, *info.Trainer.ContainerPort)
}

// ...
```

**How to Determine Default Resources**

> Currently, `torchtune` has limited support for multi-node training (but will coming soon). So, I would propose that we use 1 PyTorch node and 1 GPU by default. Users can specify `num_nodes` and `resource_per_node` in the `Trainer` field to increase PyTorch nodes and GPU number.

Reviewer: We already have multi-node support in `torchtune` isn't it @joecummings ?
Do you mean that it is not supported for all LLMs ?

Reviewer: Is there something stopping us to use `--nnodes=2` for other configs ?

Author: As is described in the [README file](https://github.com/pytorch/torchtune#supervised-finetuning-sft) of torchtune, they only officially support multinode mode in full fine-tuning.

Also, they only provide multinode config file for Llama 3.3 70B by now. We cannot guarantee that all the config items are the same when it comes to other models. For example, the FSDP config item:

In config `llama_3_3/70B_full.yaml`: https://github.com/pytorch/torchtune/blob/0e8f840e4a34d2fdb629926341582a9d23f54c2c/recipes/configs/llama3_3/70B_full.yaml#L78 

In config `llama_3_3/70B_full_multinode`: https://github.com/pytorch/torchtune/blob/0e8f840e4a34d2fdb629926341582a9d23f54c2c/recipes/configs/llama3_3/70B_full_multinode.yaml#L64

This problem is also described in the [official guide](https://pytorch.org/torchtune/main/tutorials/multinode.html#future-development). They haven't supported 2D parallelism by now.

### Modify the `train` API

As we discussed in [Github](https://github.com/kubeflow/trainer/pull/2410#discussion_r1963832826), `train` API mainly executes two types of training tasks:

1. **Training with custom function/image**: A self-contained function/image that encapsulates the entire model training process.
2. **Config-driven appproach with existing Trainer**: A trainer that already includes fine-tuning logic, requiring only parameter adjustments.

So we plan to modify `train` API to:

```python
def train(
    trainer: Optional[CustomTrainer],
    fine_tuning_config: Optional[Union[TorchTuneConfig]],
    dataset_config: Optional[types.HuggingFaceDatasetConfig] = None,
    model_config: Optional[types.HuggingFaceModelInputConfig] = None,
    runtime_ref: Optional[str] = "torchtune-llm-finetuning",

> ) -> str:

Reviewer: @astefanutti @kubeflow/wg-training-leads @Electronic-Waste @deepanker13 @saileshd1402 @franciscojavierarceo @seanlaii  What do you think about this API design ?

Reviewer: Maybe it gives us opportunity in the future to support more framework-specific trainers like TorchTrainer, TorchXLATrainer, DeepSpeedTrainer, etc.
Which will help user to appropriately configure model, dataset (assign devices and configure distributed backend).

Author: > What do you think about this API design ?

LGTM.

We natively support all `recipe` and `config` supported by `torchtune`, since `torchtune` has already provided us with default `config`. We just cannot mutate them if we do not support the corresponding mutation config.

### Propagate `torchtune` settings with SDK

#### Option1: Allow users to specify `recipe` and `config`

| Parameters | Type | What is it? |
| - | - | - |
| recipe | str | The name of recipe in `torchtune` we choose. |
| config | str | The name of config in `torchtune` we chooose. |
| dtype | Optional[str] | The underlying data type used to represent the model and optimizer parameters. Currently, we only support `bf16` and `fp32`. |
| batch_size | Optional[int] | The number of samples processed before updating model weights. |
| epochs | Optional[int] | The number of samples processed before updating model weights. |
| loss | Optional[str] | The loss algorithm we use to fine-tune the LLM, e.g. `torchtune.modules.loss.CEWithChunkedOutputLoss` |
| peft_config | Optional[Union[LoraConfig]] | Configuration for the PEFT(Parameter-Efficient Fine-Tuning), including LoRA/QLoRA/DoRA, etc. |

> | dataset_preprocess_config | Optional[Union[InstructDataset, ChatDataset, MultimodalDataset]] | Configuration for dataset preprocessing. |

Reviewer: Since we also have `dataset_config` as part of `train()` API, how do we distinguish it ?

Reviewer: Do we have any better experience to make it clear for users when they should use it?

Maybe we could create a new dataset config: `TorchTuneDatasetConfig`, where we can define dataset properties which is specific to `torchtune`.
We can always add the `storage_uri` parameter there which allows users to configure location of dataset:
```
hf://....
s3://...
```

@joecummings Do we have any capabilities in `torchtune` to pre-process dataset outside of the main `tune run` loop ?
 For example, in `multi-node` environment dataset can be pre-processed in CPU-based machine before passing data to all Training Nodes.

cc @akshaychitneni @shravan-achar @bigsur0

Reviewer: To make the UX clear, we can put `dataset_config` under `TorchTuneConfig` as well.

Author: > Since we also have dataset_config as part of train() API, how do we distinguish it ?

1. `dataset_config` is global, in charge of downloading dataset. 
2. `dataset_preprocess_config` is limited to `torchtune` only, doing the preprocess work with the help of built in Dataset Class support of `torchtune`

It might not be a good idea to put `dataset_config` into `TorchtuneConfig` since their scopes are different.

Author: Currently, the data preprocessing is coupled with recipe/config:

https://github.com/pytorch/torchtune/blob/0e8f840e4a34d2fdb629926341582a9d23f54c2c/recipes/configs/llama3_1/70B_lora.yaml#L40-L46

So maybe we have no alternatives but to preprocess them on the GPU nodes if we do not want to reinvent the wheel.

```python
# TorchtuneConfig DataClass
@dataclass
class TorchtuneConfig:
    recipe: str

>     config: str

Reviewer: As we discussed in Slack, we might want to create Runtime for each Model/Recipe.
Which means, user can select the appropriate runtime based on it.

Reviewer: But if we know what config we need to fetch, can we mutate values based on user configuration ?
Should we have list of supported config for every supported LLM in Kubeflow SDK ?
I want to design UX when user doesn't need to know exact recipe and config they should use.

Author: > But if we know what config we need to fetch, can we mutate values based on user configuration ?

Yes, we can. As I mentioned in the KEP, we just need to append CLI args to `tune run` command.

> Should we have list of supported config for every supported LLM in Kubeflow SDK ?
I want to design UX when user doesn't need to know exact recipe and config they should use to fine-tune LLM.

That's true. It's what torchtune did in their `tune ls`: https://github.com/pytorch/torchtune/blob/0e8f840e4a34d2fdb629926341582a9d23f54c2c/torchtune/_cli/ls.py#L45-L64

We can also create a CLI command like `kubeflow tune ls` (https://github.com/kubeflow/community/pull/823) to allow users to choose their recipe and config. It's easy for users to choose from them, and achieve almost the same UX as creating ~100 ClusterTrainingRuntime, which also requires much time from users to choose their ClusterTrainingRuntime.

    dtype: Optional[str] = None
    batch_size: Optional[int] = None
    epochs: Optional[int] = None
    loss: Optional[str] = None
    peft_config: Optional[Union[LoraConfig]] = None
    dataset_preprocess_config: Optional[
        Union[InstructDataset, ChatDataset, MultimodalDataset],
    ] = None

```

In this way, we just need to create one `ClusterTrainingRuntime` for LLM fine-tuning with `torchtune`. Like we showed in the [example](#compliment-torch-plugin). And we will put this `ClusterTrainingRuntime` under:

```
manifests/
|-- base/
|   |-- runtimes/
|   |   |-- pretraining/
|   |   |   |-- kustomization.yaml
|   |   |   |-- mpi_distributed.yaml                    # MPI Distributed Runtime
|   |   |   |-- torch_distributed.yaml                  # PyTorch Distributed Runtime

> |   |   |-- posttraining/

Reviewer: This needs to be re-consider according to: https://github.com/kubeflow/trainer/issues/2430
@Electronic-Waste What do you think about this folder structure:
```
manifests/
|-- base/
|   |-- runtimes/
|   |   |-- kustomization.yaml
|   |   |-- mpi_distributed.yaml                    # MPI Distributed Runtime
|   |   |-- torch_distributed.yaml                  # PyTorch Distributed Runtime
|   |   |-- torchtune/
|   |   |   |-- kustomization.yaml
|   |   |   |-- torchtune_llm_finetuning.yaml           # Torchtune LLM Fine-tuning Runtime
|   |-- crds/
```

Maybe as a label to apply for these runtimes, we should add this:
```
trainer.runtime.kubeflow.org/type: custom-trainer
trainer.runtime.kubeflow.org/phase: any

trainer.runtime.kubeflow.org/type: torchtune
trainer.runtime.kubeflow.org/phase: post-training
```

WDYT @Electronic-Waste @astefanutti @kubeflow/wg-training-leads @franciscojavierarceo ?

Author: Maybe:

```
manifests/
|-- base/
|   |-- runtimes/
|   |   |-- kustomization.yaml
|   |   |-- mpi_distributed.yaml                    # MPI Distributed Runtime
|   |   |-- torch_distributed.yaml                  # PyTorch Distributed Runtime
|   |   |-- torchtune_llm_finetuning.yaml           # Torchtune LLM Fine-tuning Runtime
|   |-- crds/
```

is better? Since we only use torchtune for LLM fine-tuning.

Reviewer 3: +1 @Electronic-Waste

Reviewer: >Since we only use torchtune for LLM fine-tuning, it might be unnecessary to open a new dir for it.

Let's first of all agree on what runtimes we want to deploy: https://github.com/kubeflow/trainer/pull/2410#discussion_r1969962609.

Also we need to create directories for each LLMs and  `ClusterTrainingRuntime`s for each config file, which might be too complex for us to maintain (~100 `ClusterTrainingRuntimes`).

```
manifests/
|-- base/
|   |-- runtimes/
|   |   |-- pretraining/
|   |   |   |-- kustomization.yaml
|   |   |   |-- mpi_distributed.yaml                    # MPI Distributed Runtime
|   |   |   |-- torch_distributed.yaml                  # PyTorch Distributed Runtime
|   |   |-- posttraining/
|   |   |   |-- kustomization.yaml
|   |   |   |-- llama2/
|   |   |   |   |-- kustomization.yaml
|   |   |   |   |-- 13B_full_torchtune.yaml             # Torchtune Full Fine-tuning Runtime for Llama2 13B
|   |   |   |   |-- . . .                               # (+10 ClusterTrainingRuntimes)
|   |   |   |-- llama3/
|   |   |   |   |-- kustomization.yaml
|   |   |   |   |-- 70B_lora_torchtune.yaml             # Torchtune LoRA Fine-tuning Runtime for Llama3 70B
|   |   |   |   |-- . . .                               # (+10 ClusterTrainingRuntimes)
|   |   |   |-- . . .
|   |-- crds/
|   |-- . . .
|-- overlays/
|-- third-party/
```

### Support some common PEFT mechanisms

> We need to support some common PEFT mechnisms like LoRA, QLoRA, DoRA to allow users to optimize the memory usage when they are fine-tuning the LLMs. This is crucial for users who have limited resources and want to fine-tune their model at the minimum cost.

Reviewer 3: I think implementing together makes sense. I also think we should probably make it extensible (maybe via `kwargs`) since PEFT algorithms seem to continue to evolve and I doubt that'll end soon (as evident in page 2 on the PEFT survey paper: https://arxiv.org/pdf/2312.12148).

> We need to support some common PEFT mechnisms like LoRA, QLoRA, DoRA to allow users to optimize the memory usage when they are fine-tuning the LLMs. This is crucial for users who have limited resources and want to fine-tune their model at the minimum cost.

Reviewer 3: Actually realized that GRPO and having a teacher model may be included in this.

We may want to look at the DeepSeek implementation a bit. HuggingFace has already implemented something: https://huggingface.co/docs/trl/main/en/grpo_trainer

    ),
    runtime_ref=llm_runtime,
)
```

## Design Details

### Config Override

As is shown in the official guides, we can pass distributed arguments to `torchtune` and override some parameters in the config, like:

```bash
# Note that tune run only supports launching distributed runs by passing through arguments preceding the recipe directly to torchrun.
# Official document: https://pytorch.org/torchtune/main/tune_cli.html#run-a-recipe
# Proof from source code: https://github.com/pytorch/torchtune/blob/75965d4281b9b76c454630d015221b9933c77bf3/torchtune/_cli/run.py#L113-L118
tune run --nnodes=1 --nproc-per-node=4 lora_finetune_distributed \
    --config llama3/8B_lora \
    model.lora_attn_modules=[q_proj,k_proj,v_proj] \
    model.apply_lora_to_mlp=True \
    model.lora_rank=64 \
    model.lora_alpha=128

```

The CLI-based way allows us to mutate these parameters by overriding the `command` and `args` fields in the Trainer Node. By specifying override parameters in the TrainJob `.spec.trainer.command` and `.spec.trainer.args` fields, we can easily mutate these parameters, just like:

```YAML
apiVersion: kubeflow.org/v2alpha1
kind: ClusterTrainingRuntime
metadata:

>   name: torchtune-llm-finetuning

Reviewer: Should you refactor this CTR and TrainJob according to our recent structure ?

Author: Yeah, of course!

  labels:

>     training.kubeflow.org/phase: post-training

Reviewer: Can we update these labels according to our discussion here: https://github.com/kubeflow/trainer/pull/2410#discussion_r1970006713 if they look good to you.
```
trainer.runtime.kubeflow.org/type: torchtune
trainer.runtime.kubeflow.org/phase: post-training
```

Author: Sure! I'll update them soon:)

      - lora_finetune_single_device
      - --config
      - llama3_2/1B_lora
      - model.lora_attn_modules=[q_proj,k_proj,v_proj]
      - model.apply_lora_to_mlp=True
      - model.lora_rank=64
      - model.lora_alpha=128
    resourcesPerNode:
      requests:
        nvidia.com/gpu: 2
    numProcPerNode: 2
...
```

The trainer image will be built with packages related to PyTorch, CUDA, Torchtune. And we will maintain its Dockerfile under `cmd/trainers/torchtune/`.

### Complement `torch` Plugin

**How to perform mutation in the runtime plugin**

We need to modify the exsiting torch plugin to handle config override for `torchtune`, since currently torch plugin passes the distributed arguments by environment variables that begins with `PET_`, which is not allowed by `torchtune`. However, `torchtune` can share the same ML Policy with `torchrun` because `torchtune` is fully compatible with these distributed parameters.

We will add CLI-based parameters mutation to `pkg/runtime/framework/torch/torch.go` while still supporting env-based parameters mutation with `PET_`:

```go
func (t *Torch) EnforceMLPolicy(info *runtime.Info, trainJob *trainer.TrainJob) error {
    // ...

>     if !trainJob.IsLaunchedByTorchtune() {

Reviewer: ```suggestion
    if !trainJob.IsLaunchedByTorchTune() {
```

        // Original env-based parameters mutation
        // Update envs for Info object.
        // Add PyTorch distributed "PET_" values for torchrun
        // TBA
    } else {
        // CLI-based parameters mutation
        // Add PyTorch distributed values to Info.Args,
        // which will be inserted in front of Args of Trainer Node.
        // TBA
    }
    // ...
}
```

> Related changes:

Reviewer: That might need to be changed according to the @astefanutti's changes for SSA: https://github.com/kubeflow/trainer/blob/master/pkg/runtime/framework/plugins/torch/torch.go#L58

Author: I'll change then soon:)

}

// Update the Trainer container port.
if info.Trainer.ContainerPort != nil {
    b.Spec.ReplicatedJobs[i].Template.Spec.Template.Spec.Containers[j].Ports = append(
        b.Spec.ReplicatedJobs[i].Template.Spec.Template.Spec.Containers[j].Ports, *info.Trainer.ContainerPort)
}

// ...
```

**How to Determine Default Resources**

Currently, `torchtune` has limited support for multi-node training (but will coming soon). So, I would propose that we use 1 PyTorch node and 1 GPU by default. Users can specify `num_nodes` and `resource_per_node` in the `Trainer` field to increase PyTorch nodes and GPU number.

### Modify the `train` API

As we discussed in [Github](https://github.com/kubeflow/trainer/pull/2410#discussion_r1963832826), `train` API mainly executes two types of training tasks:

1. **Training with custom function/image**: A self-contained function/image that encapsulates the entire model training process.
2. **Config-driven appproach with existing Trainer**: A trainer that already includes fine-tuning logic, requiring only parameter adjustments.

So we plan to modify `train` API to:

```python
def train(

>     trainer: Optional[CustomTrainer],
>     fine_tuning_config: Optional[Union[TorchTuneConfig]],

Reviewer: ```suggestion
    trainer: Optional[CustomTrainer] = None,
    fine_tuning_config: Optional[Union[TorchTuneConfig]] = None,
```

    dataset_config: Optional[types.HuggingFaceDatasetConfig] = None,
    model_config: Optional[types.HuggingFaceModelInputConfig] = None,
    runtime_ref: Optional[str] = None,
) -> str:
    pass

@dataclass
class CustomTrainer:
    func: Optional[Callable] = None
    func_args: Optional[Dict] = None
    packages_to_install: Optional[List[str]] = None
    pip_index_url: str = constants.DEFAULT_PIP_INDEX_URL
    num_nodes: Optional[int] = None
    resources_per_node: Optional[Dict] = None

```

`trainer` defines the parameters for Type 1 tasks, and will add support for custom function in the initial stage.

`fine_tuning_config` defines the parameters for LLM fine-tuning task in Type 2, and will add support for fine-tuning with `torchtune` in the early stage.

We natively support all `recipe` and `config` supported by `torchtune`, since `torchtune` has already provided us with default `config`. We just cannot mutate them if we do not support the corresponding mutation config.

### Propagate `torchtune` settings with SDK

> To provide a better user experience, we need to offer a simple SDK that allows users to easily modify config files. So, we introduce `TorchtuneConfig` dataclass and create a map from (`TorchtuneConfig`, `num_nodes`, `runtime_ref`) to dedicated `recipe` and `config`. Then, users can fine-tune their LLMs without any knowledge about `torchtune`.

Reviewer: ```suggestion
To provide a better user experience, we need to offer a simple SDK that allows users to easily modify config files. So, we introduce `TorchTuneConfig` dataclass and create a map from (`TorchTuneConfig`, `num_nodes`, `runtime_ref`) to dedicated `recipe` and `config`. Then, users can fine-tune their LLMs without any knowledge about `torchtune`.
```

```python
# By default we can fine-tune models without any additional configurations from users
TrainerClient().train(
  runtime_ref="torchtune-llama-3.3-70b"
)
```

> #### `TorchtuneConfig` API Design

Reviewer: ```suggestion
#### `TorchTuneConfig` API Design
```

| Parameters | Type | What is it? |
| - | - | - |
| dtype | Optional[str] | The underlying data type used to represent the model and optimizer parameters. Currently, we only support `bf16` and `fp32`. |
| batch_size | Optional[int] | The number of samples processed before updating model weights. |
| epochs | Optional[int] | The number of samples processed before updating model weights. |
| loss | Optional[str] | The loss algorithm we use to fine-tune the LLM, e.g. `torchtune.modules.loss.CEWithChunkedOutputLoss` |
| peft_config | Optional[Union[LoraConfig]] | Configuration for the PEFT(Parameter-Efficient Fine-Tuning), including LoRA/QLoRA/DoRA, etc. |
| dataset_preprocess_config | Optional[Union[InstructDataset, ChatDataset, MultimodalDataset]] | Configuration for dataset preprocessing. |
| num_nodes | Optional[int] | The number of PyTorch Nodes in training |
| resource_per_node | Optional[Dict] | The resource for each PyTorch Node |

```python

> # TorchtuneConfig DataClass
> @dataclass
> class TorchtuneConfig:

Reviewer: ```suggestion
# TorchTuneConfig DataClass
@dataclass
class TorchTuneConfig:
```

    dtype: Optional[str] = None
    batch_size: Optional[int] = None
    epochs: Optional[int] = None
    loss: Optional[str] = None
    peft_config: Optional[Union[LoraConfig]] = None
    dataset_preprocess_config: Optional[
        Union[InstructDataset, ChatDataset, MultimodalDataset],
    ] = None
    num_nodes: Optional[int] = None,
    resources_per_node: Optional[Dict] = None,

```

> #### Create Map from `TorchtuneConfig` to specific recipes and configs
> 
> As we mentioned above, we will create a map from (`TorchtuneConfig`, `num_nodes`, `runtime_ref`) to dedicated `recipe` and `config`. This will allow users to fine-tune their LLMs without knowing about `torchtune`'s recipes and configs.

Reviewer: ```suggestion
#### Create Map from `TorchTuneConfig` to specific recipes and configs

As we mentioned above, we will create a map from (`TorchTuneConfig`, `num_nodes`, `runtime_ref`) to dedicated `recipe` and `config`. This will allow users to fine-tune their LLMs without knowing about `torchtune`'s recipes and configs.
```

**How to Select `recipe`**

1. `peft_config == None` && `num_nodes == 1` && `nproc_per_node == 1`: Use `full_finetune_single_device`.
2. `peft_config == None` && (`num_nodes >= 1` || `nproc_per_node > 1`): Use `full_finetune_distributed`.
3. `type(peft_confg) == LoraConfig` && `num_nodes == 1`: Use `lora_finetune_single_device`.
4. `type(peft_confg) == LoraConfig` && `num_nodes > 1`: Use `lora_finetune_distributed`
5. TBA if we want to support more fine-tuning techniques.

**How to Select `config`**

We will create one `ClusterTrainingRuntime` for one model. In this way, we can extract the model info in `runtime_ref`, and select corresponding config file according to the `recipe`.

### Maintain ClusterTrainingRuntimes in Manifests

According to our design in [this section](#propagate-torchtune-settings-with-sdk), we will create one `ClusterTrainingRuntime` for one model. We can maintain these `ClusterTraingRuntimes` under `manifests/base/runtimes`:

```
manifests/
|-- base/
|   |-- runtimes/
|   |   |-- kustomization.yaml
|   |   |-- mpi_distributed.yaml                    # MPI Distributed Runtime
|   |   |-- torch_distributed.yaml                  # PyTorch Distributed Runtime

> |   |   |-- torchtune/                              # Torchtune Fine-tuning Runtimes

Reviewer: I think, in the first iteration, we can just support 1 or 2 LLMs and gather the initial user feedback.
For example, I don't know if we should support Llama 3.1 and Llama 3.2 together.

Author: That's a good idea! Llama 3.1 and Llama 3.2 are widely used in the world. In that case, we could first support Llama models family (3.1, 3.2, 3,3) so that users can have more choices.

Reviewer: @Electronic-Waste For diversity please can we keep Llama 3.2, Llama 3.1, Qwen, and Gemma as well ?
I think, we can add it to your proposal folder structure, but implement the runtime after we support at least one LLM.

Author: For sure, I'll add them back:)

|   |   |   |-- kustomization.yaml
|   |   |   |-- gemma2/

> |   |   |   |   |-- torchtune_gemma2_2B_finetuning.yaml
> |   |   |   |   |-- torchtune_gemma2_7B_finetuning.yaml
> |   |   |   |   |-- torchtune_gemma2_27B_finetuning.yaml

Reviewer: I think, `torchtune` is redundant in the name, since we already place them under `torchtune/` directory.
WDYT @Electronic-Waste ?
```suggestion
|   |   |   |   |-- 2B_finetuning.yaml
|   |   |   |   |-- 7B_finetuning.yaml
|   |   |   |   |-- 27B_finetuning.yaml
```

Reviewer: I guess, torchtune does the same: https://github.com/pytorch/torchtune/tree/main/recipes/configs/gemma2

Author: SGTM.

      - name: STORAGE_URI
        value: hf://tatsu-lab/alpaca
    volumeMounts:
      - mountPath: /workspace/dataset
        name: dataset-initializer
volumes:
  - name: dataset-initializer
    persistentVolumeClaim:
      claimName: dataset-initializer

```

```yaml
# Model Initialization
containers:
  - name: model-initializer
    image: docker.io/kubeflow/model-initializer
    env:
      - name: STORAGE_URI
        value: hf://meta-llama/Llama-2-7b
    volumeMounts:
      - mountPath: /workspace/model
        name: model-initializer
volumes:
  - name: model-initializer
    persistentVolumeClaim:
      claimName: model-initializer

```

As for the [model exporter](https://github.com/kubeflow/trainer/issues/2245), we haven't implemented it yet in Kubeflow Trainer. But we'll use it as the exporter for our fine-tuned LLMs once it's ready. For example:

```yaml
# Model Exporting
containers:
  - name: model-exporter
    image: docker.io/kubeflow/model-exporter
    volumeMounts:
      - mountPath: /workspace/adapters
        name: model-exporter
volumes:
  - name: model-exporter
    persistentVolumeClaim:
      claimName: model-exporter

```

### Dataset Preprocess / Tokenizer

> `torchtune` has supported several types of [dataset classes](https://pytorch.org/torchtune/main/basics/datasets_overview.html), including Instruct, Chat, Multimodal, which will preprocess dataests for us automatically. We just need to configure it in the SDK and override it to the config:

Reviewer: ```suggestion
`torchtune` has supported several types of [dataset classes](https://pytorch.org/torchtune/main/basics/datasets_overview.html), including Instruct, Chat, Multimodal, which will preprocess dataset for us automatically. We just need to configure it in the SDK and override it to the config:
```

**Instruct Dataset**

```python
@datasetclass
class InstructDataset:

>     source: Optional[str] = None

Reviewer: What does source mean here ?
Given that we are going to use our dataset initializer, should the source always be equal to:
`/workspace/dataset` ?

Author: Oh, that's true. I'll delete it.

Author: `source` is the data source type if we use local datasets downloaded by dataset initializer, such as `json`: https://pytorch.org/torchtune/main/basics/chat_datasets.html#loading-local-and-remote-chat-datasets

Reviewer: I can see that for the HF datasets the source is equal to repo name: https://pytorch.org/torchtune/main/basics/chat_datasets.html#loading-chat-datasets-from-hugging-face.

Can we just point source to the `/workspace/dataset` directory all the time ?

Author: > I can see that for the HF datasets the source is equal to repo name

Yeah. That's because `torchtune` supports downloading datasets from HF by specifying the `source`, just like our data initializer.

> Can we just point source to the /workspace/dataset directory all the time ?

Since we will always use our initializer for downloading datasets, we need to specify `source` to `json` to indicate that we disable downloading from HF and use our local dataset. We'll use `data_files` to specify the path to datasets.

```python
from torchtune.models.gemma import gemma_tokenizer
from torchtune.datasets import chat_dataset

g_tokenizer = gemma_tokenizer("/tmp/gemma-7b/tokenizer.model")
ds = chat_dataset(
    tokenizer=g_tokenizer,
    source="json",
    conversation_column="conversations",
    conversation_style="sharegpt",
    data_files="data/my_data.json",
    split="train",
)
```

As an alternative, we can remove `data_files` and add `data_file_name`: `data_files` = `/workspace/dataset/` + `data_file_name`. But we can discuss it further in the following PRs:)

Reviewer: Make sense, I guess it depends on the data format (e.g. json, parquet, sql) as described here: https://huggingface.co/docs/datasets/main/en/loading#local-and-remote-files.

Let's discuss in the future if we should expose this information to the user or we can intelligently understand it during our orchestration (e.g. check the actual files we download from HF).

Author: SGTM!

        // which will be inserted in front of Args of Trainer Node.
        // TBA
    }
    // ...
}
```

Related changes:

1. Add `Args` fields to Info (`pkg/runtime/runtime.go`).

```go
// We use SSA to reconcile TrainJob now.
// Ref: https://github.com/kubeflow/trainer/pull/2431
type Trainer struct {
    NumNodes       *int32
    NumProcPerNode string
    Args           []string
    Env            []corev1ac.EnvVarApplyConfiguration
    ContainerPort  *corev1ac.ContainerPortApplyConfiguration
    Volumes        []corev1ac.VolumeApplyConfiguration
    VolumeMounts   []corev1ac.VolumeMountApplyConfiguration
}
```

2. Insert `Info.Args` in front of Args of Trainer Node (`pkg/runtime/framework/plugins/jobest/builder.go`)

```go
// ...

// Update the Trainer args.
if args := info.Trainer.Args; args != nil {
    apply.UpsertArg(b.Spec.ReplicatedJobs[i].Template.Spec.Template.Spec.Containers[j].Args, args)
}

// Update the Trainer container port.
if port := info.Trainer.ContainerPort; port != nil {
    apply.UpsertPort(&b.Spec.ReplicatedJobs[i].Template.Spec.Template.Spec.Containers[j].Ports, port)
}

// ...
```

**How to Determine Default Resources**

> Currently, `torchtune` has limited support for multi-node training (but will coming soon). So, I would propose that we use 1 PyTorch node and 1 GPU by default. Users can specify `num_nodes` and `resource_per_node` in the `Trainer` field to increase PyTorch nodes and GPU number.

Reviewer 3: ```suggestion
Currently, `torchtune` has limited support for multi-node training (but support is coming soon). So, I would propose that we use 1 PyTorch node and 1 GPU by default. Users can specify `num_nodes` and `resource_per_node` in the `Trainer` field to increase PyTorch nodes and GPU number.
```

### Modify the `train` API

As we discussed in [Github](https://github.com/kubeflow/trainer/pull/2410#discussion_r1963832826), `train` API mainly executes two types of training tasks:

> 1. **Training with custom function/image**: A self-contained function/image that encapsulates the entire model training process.

Reviewer 3: ```suggestion
1. **Type 1: Training with custom function/image**: A self-contained function/image that encapsulates the entire model training process.
```

> 2. **Config-driven appproach with existing Trainer**: A trainer that already includes fine-tuning logic, requiring only parameter adjustments.

Reviewer 3: ```suggestion
2. **Type 2: Config-driven approach with existing Trainer**: A trainer that already includes fine-tuning logic, requiring only parameter adjustments.
```

So we plan to modify `train` API to:

```python
def train(
    trainer: Optional[CustomTrainer] = None,
    fine_tuning_config: Optional[Union[TorchTuneConfig]] = None,
    dataset_config: Optional[types.HuggingFaceDatasetConfig] = None,
    model_config: Optional[types.HuggingFaceModelInputConfig] = None,
    runtime_ref: Optional[str] = None,
) -> str:
    pass

@dataclass
class CustomTrainer:
    func: Optional[Callable] = None
    func_args: Optional[Dict] = None
    packages_to_install: Optional[List[str]] = None
    pip_index_url: str = constants.DEFAULT_PIP_INDEX_URL
    num_nodes: Optional[int] = None
    resources_per_node: Optional[Dict] = None

```

> `trainer` defines the parameters for Type 1 tasks, and will add support for custom function in the initial stage.

Reviewer 3: I had to think a bit to understand this was referring to the (1) mentioned above, so I made a suggestion to make it more obvious for the aloof like me. :)

Author: SGTM! Thanks for your precious suggestions!

      - name: STORAGE_URI
        value: hf://tatsu-lab/alpaca
    volumeMounts:
      - mountPath: /workspace/dataset
        name: dataset-initializer
volumes:
  - name: dataset-initializer
    persistentVolumeClaim:
      claimName: dataset-initializer

```

```yaml
# Model Initialization
containers:
  - name: model-initializer
    image: docker.io/kubeflow/model-initializer
    env:
      - name: STORAGE_URI
        value: hf://meta-llama/Llama-2-7b
    volumeMounts:
      - mountPath: /workspace/model
        name: model-initializer
volumes:
  - name: model-initializer
    persistentVolumeClaim:
      claimName: model-initializer

```

As for the [model exporter](https://github.com/kubeflow/trainer/issues/2245), we haven't implemented it yet in Kubeflow Trainer. But we'll use it as the exporter for our fine-tuned LLMs once it's ready. For example:

```yaml
# Model Exporting
containers:
  - name: model-exporter
    image: docker.io/kubeflow/model-exporter
    volumeMounts:
      - mountPath: /workspace/adapters
        name: model-exporter
volumes:
  - name: model-exporter
    persistentVolumeClaim:
      claimName: model-exporter

```

### Dataset Preprocess / Tokenizer

> `torchtune` has supported several types of [dataset classes](https://pytorch.org/torchtune/main/basics/datasets_overview.html), including Instruct, Chat, Multimodal, which will preprocess the dataest for us automatically. We just need to configure it in the SDK and override it to the config:

Reviewer 3: nice

```python
# TorchTuneConfig DataClass
@dataclass
class TorchTuneConfig:
    dtype: Optional[str] = None
    batch_size: Optional[int] = None
    epochs: Optional[int] = None
    loss: Optional[str] = None
    peft_config: Optional[Union[LoraConfig]] = None
    dataset_preprocess_config: Optional[
        Union[InstructDataset, ChatDataset, MultimodalDataset],
    ] = None
    num_nodes: Optional[int] = None,
    resources_per_node: Optional[Dict] = None,

```

### Complement `torch` Plugin

#### Perform Mutation in `torch` Plugin

We need to modify the exsiting torch plugin to handle config override for `torchtune`, since currently torch plugin passes the distributed arguments by environment variables that begins with `PET_`, which is not allowed by `torchtune`. However, `torchtune` can share the same ML Policy with `torchrun` because `torchtune` is fully compatible with these distributed parameters.

We will add CLI-based parameters mutation to `pkg/runtime/framework/torch/torch.go` while still supporting env-based parameters mutation with `PET_`:

```go
func (t *Torch) EnforceMLPolicy(info *runtime.Info, trainJob *trainer.TrainJob) error {
    // ...

>     if !trainJob.IsLaunchedByTorchTune() {

Reviewer: Separate plugin might be a good idea, I agree, since we can set custom validation + Info object creation based on it.
However, we need to re-use some assets from the `torch` plugin (e.g. validation of `numProcPerNode`, adding `torchrun` env variable).
@tenzen-y Any thoughts ? Should we execute `TorchTune` + `Torch` plugins together when we run these LLM blueprints ?

>     if !trainJob.IsLaunchedByTorchTune() {

Reviewer 5: The reason why I postponed the separate plugin decision is unclear what we will do for torchtune in the plugin.

So, I would recommend 

Clear all things for actual mechanism for torchtune in plugin like what the plugin do in this documents.

or just postpone it to impl PR.

It might be better to postpone it to impl PR since it looks like Shao currently does not have what exactly should do in plugin. In this design, the mechanism description sounds a slightly lack of information.

>     if !trainJob.IsLaunchedByTorchTune() {

Reviewer 5: I meant based on this design, I can not decide which solution (torch plugin or torchtune plugin) would be better.

For sure, impl has incorrect direction, I will propose another solution in the PR.

>     if !trainJob.IsLaunchedByTorchTune() {

Reviewer: Sounds good, we can follow-up when we start the implementation

>     if !trainJob.IsLaunchedByTorchTune() {

Author: Yeah, I agree that we can postpone the detailed design to the implementation PR:)

        // Original env-based parameters mutation
        // Update envs for Info object.
        // Add PyTorch distributed "PET_" values for torchrun.
        // TBA
    } else {
        // CLI-based parameters mutation
        // Add PyTorch distributed values to Info.Args,
        // which will be inserted in front of Args of Trainer Node.
        // TBA
    }
    // ...
}
```

Related changes:

1. Add `Args` fields to Info (`pkg/runtime/runtime.go`).

```go
// We use SSA to reconcile TrainJob now.
// Ref: https://github.com/kubeflow/trainer/pull/2431
type Trainer struct {
    NumNodes       *int32
    NumProcPerNode string
    Args           []string
    Env            []corev1ac.EnvVarApplyConfiguration
    ContainerPort  *corev1ac.ContainerPortApplyConfiguration
    Volumes        []corev1ac.VolumeApplyConfiguration
    VolumeMounts   []corev1ac.VolumeMountApplyConfiguration
}

> ```

Author: Sure, I'll remove it.

> ```

Author: > This has been removed to consider all roles except for Trainer. Instead of this, you can add args to runtime.Container

Do you mean that I should create another struct `Container` in `pkg/runtime/runtime.go`? Can you elaborate a bit? Thanks!

> ```

Reviewer 5: this one https://github.com/kubeflow/trainer/blob/590f258047f1ad0e128bddb050b3e3bceb800ad2/pkg/runtime/runtime.go#L74

1. The ClusterTrainingRuntime referenced by `runtime_ref` exists in the control plane.
2. TBA...

#### Determine Default Resources

Currently, `torchtune` has limited support for multi-node training (but support is coming soon). So, we'd better use 1 PyTorch node and 1 GPU by default. Users can specify `num_nodes` and `resource_per_node` in the `Trainer` field to increase PyTorch nodes and GPU number.

### Maintain ClusterTrainingRuntimes in Manifests

According to our design in [this section](#propagate-torchtune-settings-with-sdk), we will create one `ClusterTrainingRuntime` for one model. We can maintain these `ClusterTraingRuntimes` under `manifests/base/runtimes`:

```
manifests/
|-- base/
|   |-- runtimes/
|   |   |-- kustomization.yaml
|   |   |-- mpi_distributed.yaml                    # MPI Distributed Runtime
|   |   |-- torch_distributed.yaml                  # PyTorch Distributed Runtime
|   |   |-- torchtune/                              # TorchTune Fine-tuning Runtimes
|   |   |   |-- kustomization.yaml
|   |   |   |-- gemma2/

> |   |   |   |   |-- 2B_finetuning.yaml
> |   |   |   |   |-- 7B_finetuning.yaml
> |   |   |   |   |-- 27B_finetuning.yaml

Reviewer: Do you prefer this or this to be consistent with runtime name: `torchtune-gemma2-2b`
?
```suggestion
|   |   |   |   |-- gemma_2B.yaml
|   |   |   |   |-- gemma_7B.yaml
|   |   |   |   |-- gemma_27B.yaml
```

Author: @andreyvelich Since we've already created `torchtune/gemma2`, is it a bit of redundant to add a `torchtune_` prefix to the CTR file? I remember `torchtune` did the same like: https://github.com/pytorch/torchtune/tree/main/recipes/configs/gemma2

Reviewer: Yeah, I know that we already have it under the folder and `torchtune` does the same.
I just feel that keeping the model name in the file name could help us to avoid naming conflicts (e.g. `3B_finetuning.yaml` ).
For instance, if cluster admins don't want to have the same directory structure for `torchtune` runtimes (as we do upstream).

Author: SGTM.

    ] = None
    num_nodes: Optional[int] = None,
    resources_per_node: Optional[Dict] = None,

```

#### Create Map from `TorchTuneConfig` to specific recipes and configs

As we mentioned above, we will create a map from (`TorchTuneConfig`, `num_nodes`, `runtime_ref`) to dedicated `recipe` and `config`. This will allow users to fine-tune their LLMs without knowing about `torchtune`'s recipes and configs.

- How to Select `recipe`

1. `peft_config == None` && `num_nodes == 1` && `nproc_per_node == 1`: Use `full_finetune_single_device`.
2. `peft_config == None` && (`num_nodes >= 1` || `nproc_per_node > 1`): Use `full_finetune_distributed`.
3. `type(peft_confg) == LoraConfig` && `nproc_per_node == 1`: Use `lora_finetune_single_device`.
4. `type(peft_confg) == LoraConfig` && `nproc_per_node > 1`: Use `lora_finetune_distributed`.
5. TBA if we want to support more fine-tuning techniques.

- How to Select `config`

We will create one `ClusterTrainingRuntime` for one model. In this way, we can extract the model info in `runtime_ref`, and select corresponding config file according to the `recipe`.

#### Validate Fine-Tuning Configurations

> In order to ensure the validity of the configurations propagated by SDK, we plan to add some validating requirements to the TrainJob Webhook. We'll check:

Reviewer 5: ```suggestion
In order to ensure the validity of the configurations propagated by SDK, we plan to add some validating requirements to the TrainJob Webhook. We implement validations in torch plugin [`CustomValidationPlugin`](https://github.com/kubeflow/trainer/blob/1f443729ebdb3e792d4b9cd2b242f33b0e86fe14/pkg/runtime/framework/interface.go#L34-L37) :
```

Author: SGTM.

1. Model name in `model_config` is consistent with the `runtime_ref` used for fine-tuning.
2. Model referenced by `runtime_ref` supports multi-node fine-tuning if `num_nodes > 1`.
3. Model referenced by `runtime_ref` supports full fine-tuning if using the full fine-tuning recipe.
4. Model referenced by `runtime_ref` supports LoRA fine-tuning if using the LoRA fine-tuning recipe.

> 5. TBA...

Author: > Could you mention the exact field name instead of the object name?

Sure. I'll update it soon.

> Additionally, we want to reveal what are criteria for "supports XXX fine-tuning". In the current description, there are no concreate criteria. We can identify which components / plugin should serve the validations based on criteria.

I want to express that the config file of the model for XXX fine-tuning should appear in the `tune ls`'s list, such as:

```shell
$ tune ls
RECIPE                                   CONFIG
full_finetune_single_device              llama2/7B_full_low_memory
                                         mistral/7B_full_low_memory
full_finetune_distributed                llama2/7B_full
                                         llama2/13B_full
                                         mistral/7B_full
lora_finetune_single_device              llama2/7B_lora_single_device
                                         llama2/7B_qlora_single_device
                                         mistral/7B_lora_single_device
```

I'll rephrase my expressions:)

> 5. TBA...

Reviewer: I think, the support criteria for the model is that the appropriate ClusterTrainingRuntime exists on the control plane, right ? We should check it during the validation when we compare storageURI in ModelConfig:
```
STORAGE_URI: hf://Qwen/Qwen2.5-3B
```

We can check by choosing one of these 2 ways:
1. Adding the label to identify LLM for the particular runtime:
```
trainer.runtime.kubeflow.org/type: torchtune
training.kubeflow.org/phase: post-training
training.kubeflow.org/model: qwen2.5-3B
```
2. Using regex in CTR name:
```
name: torchtune-<MODEL_NAME>-finetuning
name: torchtune-qwen2.5-3B-finetuning
```

WDYT @tenzen-y @Electronic-Waste ?

> 5. TBA...

Author: Personally, I prefer option 1, for it decouples the CTR name with the validating logic.

> 5. TBA...

Reviewer: Also, how do we like this label to identify method/framework runtime is using @franciscojavierarceo @tenzen-y @Electronic-Waste ?
```
trainer.runtime.kubeflow.org/method: torchtune
trainer.runtime.kubeflow.org/method: custom-trainer
trainer.runtime.kubeflow.org/method: ....
```

> 5. TBA...

Author: I think maybe `framework` is better, since we might support other frameworks like NeMo, HF in the future. And also, users may want to find CTRs for the specific framework, such as `mpi` and `pytorch`. It's more fine-grained compared to the definition of `torchtune` and `custom-trainer`.

From my perspective, it might be better to define it as:

```
runtime.trainer.kubeflow.org/framework: torchtune
runtime.trainer.kubeflow.org/phase: post-training
runtime.trainer.kubeflow.org/model: qwen2.5-3B
```

Because `trainer.kubeflow.org` is our API Group name. 

Or we can discard the `runtime` prefix:

```
trainer.kubeflow.org/framework: torchtune
trainer.kubeflow.org/phase: post-training
trainer.kubeflow.org/model: qwen2.5-3B
```

> 5. TBA...

Reviewer: Yeah, you are right, we should use: `runtime.trainer.kubeflow.org`.
The only concern with framework is that we can't apply it for CustomTrainer.
```
runtime.trainer.kubeflow.org/method: custom-trainer
```

However, since yesterday we discussed with @saileshd1402, that we want to use `model_config` parameter to set model, maybe we should design a new API for it, rather then re-using `list_runtimes()`:
```
list_models()
```

Any thoughts @tenzen-y ?

> 5. TBA...

Author: `CustomTrainer` is a bigger concept than `torchtune`, and it's opposite is `PreBuiltTrainer`. How about using two different labels to describe them?

```
runtime.trainer.kubeflow.org/type: pre-built-trainer
runtime.trainer.kubeflow.org/framework: torchtune
runtime.trainer.kubeflow.org/phase: post-training
runtime.trainer.kubeflow.org/model: qwen2.5-3B
```

```
runtime.trainer.kubeflow.org/type: custom-trainer
runtime.trainer.kubeflow.org/framework: pytorch
runtime.trainer.kubeflow.org/phase: pre-training
```

> 5. TBA...

Reviewer 5: I think we should add `tune` to `.spec.mlPolicy.torch` and store the parameters in it.
It would be better not to use labels and annotations as parameter store.

This looks like a sign for making metadata as garbage box

> 5. TBA...

Author: > Additionally, what is objective trainer.kubeflow.org/model label?
Why not modelConfig fields?

We'll specify the `model_config.storage_uri` to download the model, while referencing a CTR for `runtime_ref`. We need to make sure they use the same model.

> 5. TBA...

Reviewer 5: In that case, why not adding modelConfig to runtime CRD?
If we want to handle mutation and validation by controllers, we should use stuructured data.

> 5. TBA...

Author: That's true. For LLM ClusterTrainingRuntime, we may not need to change the model.

WDYT @andreyvelich

> 5. TBA...

Author: Sometimes, we may need to change the source model, such as S3 for some use cases. In this case, the path of S3 storage might not obey the naming convention, and we can't do the validation on it. Does it sound good to you to remove the validation between `runtime_ref` and `model_config.storage_uri` and let users to ensure the correctness? @andreyvelich

> 5. TBA...

Reviewer: Yes, I have the same feeling with S3 or any other model registry location.
They path might be different from the model name:
path: s3://bucket/custom-path
model: qwen2.5-1b

> 5. TBA...

Author: I'll remove the related labels. Maybe we could create a separate issue to discuss about the label and new API `list_models()`.

> 5. TBA...

Reviewer: Please let's followup on the label discussion in Slack: https://cloud-native.slack.com/archives/C0742LDFZ4K/p1741263570091899

> We will do the validation in `pkg/runtime/framework/torch/torch.go`:
> 
> ```Go
> func (t *Torch) Validate(runtimeJobTemplate client.Object, runtimeInfo *runtime.Info, oldObj, newObj *trainer.TrainJob) (admission.Warnings, field.ErrorList) {
>     // Validation logic here
> }
> ```

Reviewer 5: ```suggestion
```

It would be better not to mention such impl in design. We should describe validation criteria by natural language and mathematics.

Author: Thanks for pointing this out. I'll modify them.
